Iss830 write bop renderer

wboerdijk opened the pull request:
#830 
replaces the vispy renderer in `write_bop()` with pyrenderer. 
directly converts the `MeshObjects` into `trimesh.Trimesh` objects, so we don't need to reload anything here.
whole process itself is pretty fast (should be ~ 1s for 25 frames for a single object)
tested by generating some bop dataset scenes; results look as expected (but there's still a test regarding this right?)

cornerfarmer commented:
BlenderProcHelper result for run 494:

BlenderProcHelper has started, waiting for first results.

cornerfarmer commented:
BlenderProcHelper result for run 495:

BlenderProcHelper has started, waiting for first results.

Sebastian-Jung commented:
Since `trimesh` and `pyrender` are added as a dependency, maybe we could add it to the list of required packages for the blender python interpreter? I  guess  https://github.com/DLR-RM/BlenderProc/blob/762bd65e6f37b4c8e00f8d4e0826b4575f005fea/blenderproc/run.py#L20 is deprecated @cornerfarmer ? Maybe we should add something similar to https://github.com/DLR-RM/BlenderProc/blob/main/blenderproc/command_line.py or  we add it in the BopWriter file and in the MeshObject file?

cornerfarmer commented:
BlenderProcHelper result for run 497:

BlenderProcHelper has started, waiting for first results.

wboerdijk commented:
I've now added `pyrender` and `trimesh` to [the default config](https://github.com/wboerdijk/BlenderProc/blob/iss830_write_bop_renderer/blenderproc/python/utility/DefaultConfig.py#L49-L50), is that fine?

cornerfarmer commented:
So the problem with the changes in the `bop_object_pose_sampling` example seems to be caused by scaling the trimeshes in https://github.com/DLR-RM/BlenderProc/pull/881/files#diff-1e544e5a844abc57d917a0daa321cfc51f215aa8c39fe29b5ba8d020cae760feR140.
By calling this scaling function, the random state of numpy is changed, see https://github.com/mikedh/trimesh/blob/36c7b8e6a8a392c82bdf1c81db55c450997573a1/trimesh/transformations.py#L2262
Thats why the objects are sampled at different locations in the following iterations.

So, I think we can ignore these changes. 
Also all the are test warnings can be ignored, as they are just minor deviations caused by noise.
Except `bop_object_on_surface_sampling`, here some culling seems to go wrong. @wboerdijk could you take a look at that?

cornerfarmer commented on blenderproc/python/types/MeshObjectUtility.py:
```diff
@@ -522,6 +523,17 @@ def add_modifier(self, name: str, **kwargs):
             setattr(modifier, key, value)
 
+    def mesh_as_trimesh(self) -> Trimesh:
+        """ Returns a trimesh.Trimesh instance of the MeshObject. """
```
Docu of return value is missing

cornerfarmer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -131,14 +132,41 @@ def write_bop(output_dir: str, target_objects: Optional[List[MeshObject]] = None
+            trimesh_obj = obj.mesh_as_trimesh()
+            if m2mm:
+                trimesh_obj.apply_scale(scaling=0.001)
+            trimesh_objects[obj.get_cp('category_id')] = pyrender.Mesh.from_trimesh(mesh=trimesh_obj)
```
Is this safe? What if there are objects with the same category id?

wboerdijk replied:
i think if the objects are loaded with `load_bop_objs` then this is a unique mapping from object models (i.e., if an instance exists twice, the model exists twice too). See [here](https://github.com/DLR-RM/BlenderProc/blob/762bd65e6f37b4c8e00f8d4e0826b4575f005fea/blenderproc/python/loader/BopLoader.py#L349); and the corresponding `gt_id` in the annotations also references that ([here](https://github.com/DLR-RM/BlenderProc/blob/762bd65e6f37b4c8e00f8d4e0826b4575f005fea/blenderproc/python/writer/BopWriterUtility.py#L281))

MartinSmeyer replied:
Keep in mind that there can be multiple objects from one category id in BOP and that the bopwriter should also work without the boploader :)

MartinSmeyer replied:
It's actually fine, but could be more efficient 
https://github.com/DLR-RM/BlenderProc/pull/881#discussion_r1186861587

cornerfarmer replied:
But its not always the case that objects with the same category id have the same mesh, right?
Sometimes there are different object instances that share the same category id

MartinSmeyer replied:
> But its not always the case that objects with the same category id have the same mesh, right?
Sometimes there are different object instances that share the same category id

For BOP datasets `category_id` represents the `obj_id`, i.e. a specific instance. I think this was to make it compatible to other components such as the coco_writer.

cornerfarmer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -614,16 +655,28 @@ def calc_gt_info(chunk_dirs: List[str], dataset_objects: List[MeshObject], start
                     depth_gt = depth_gt_large[
                                ren_cy_offset:(ren_cy_offset + im_height),
                                ren_cx_offset:(ren_cx_offset + im_width)]
+                    depth_gt *= 1000.  # convert to mm
```
Shouldnt there also be an if around that or is depth always stored in mm?

wboerdijk replied:
The blenderproc generated depth is always changed to mm [here](https://github.com/DLR-RM/BlenderProc/blob/main/blenderproc/python/writer/BopWriterUtility.py#L531)

cornerfarmer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -131,14 +132,41 @@ def write_bop(output_dir: str, target_objects: Optional[List[MeshObject]] = None
+    bop_pose[:3, :3] = cam_R_m2c
+    bop_pose[:3, 3] = cam_t_m2c
+
+    # flip z-axis
+    rot = np.eye(4)
+    rot[1, 1], rot[2, 2] = -1, -1
+    return rot.dot(bop_pose)
```
Can we use here the already existing utility function, to make it more readable:
```suggestion
    return change_target_coordinate_frame_of_transformation_matrix(bop_pose, ["X", "-Y", "-Z"])
```

Needs the import:
```
from blenderproc.python.utility.MathUtility import change_target_coordinate_frame_of_transformation_matrix
```

cornerfarmer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -131,14 +132,41 @@ def write_bop(output_dir: str, target_objects: Optional[List[MeshObject]] = None
 
 
+def bop_pose_to_pyrender_coordinate_system(cam_R_m2c: np.ndarray, cam_t_m2c: np.ndarray) -> np.ndarray:
+    """ Converts an object pose in bop format to pyrender coordinate system
```
Just to make it clear that we are in the camera frame here
```suggestion
    """ Converts an object pose in bop format to pyrender camera coordinate system
```

cornerfarmer commented:
BlenderProcHelper result for run 498:

BlenderProcHelper has started, waiting for first results.

cornerfarmer commented:
BlenderProcHelper result for run 499:

BlenderProcHelper has started, waiting for first results.

wboerdijk commented:
Thanks for your feedback @cornerfarmer !
Two additional updates:
- I've fixed the issue with missing masks, this is due to meshes being non-watertight. Then, depending on the material, pyrender (and already trimesh) treats them as invisible
- I've also rewritten the scaling; now instead of converting everything to m, we render regardless of unit. When rendering in mm, the depth will directly have the desired resolution; otherwise we convert it to mm [here](https://github.com/wboerdijk/BlenderProc/blob/iss830_write_bop_renderer/blenderproc/python/writer/BopWriterUtility.py#L579-L581). Results shouldn't change much; I've tested it locally with one object and the depth difference is below 1 mm which should be sufficient for our application. To get this to work we increase depth distance to 100000 (i.e., in mm this would be 100m) [here](https://github.com/wboerdijk/BlenderProc/blob/iss830_write_bop_renderer/blenderproc/python/writer/BopWriterUtility.py#L556).
- The urdf example now works again; similar to [bop poses](https://github.com/wboerdijk/BlenderProc/blob/b27dcf05d4307efc94d37028fd7a73a59d1be06c/blenderproc/python/writer/BopWriterUtility.py#L295-L299) it only supports [the first visual](https://github.com/wboerdijk/BlenderProc/blob/b27dcf05d4307efc94d37028fd7a73a59d1be06c/blenderproc/python/writer/BopWriterUtility.py#L139-L143), if existing

There still is an issue with the calculation of the visibility masks which happens based on depth image comparisons ([code](https://github.com/wboerdijk/BlenderProc/blob/c5607a83b290769b7b8099f0be499590fc1cff66/blenderproc/python/writer/BopWriterUtility.py#L577-L591)), where there are occasionally small pixel-wise artifacts (also visible in the rendered images, e.g. [this one](https://camo.githubusercontent.com/b9fab24ff62d10355e2b33bea4e63c1091046163ebab60fa3b8db048ee397f30/68747470733a2f2f6c6976652e737461746963666c69636b722e636f6d2f36353533352f35323836363732353431345f323966623332323039362e6a7067)). This only happens at borders of objects, and should be due to different alias handling in blender and pyrender I assume? I tried playing around with disabling anti-aliasing (e.g. [this suggestion](https://github.com/mmatl/pyrender/pull/175/files)), but similar artifacts remain. Maybe you have any additional suggestions @MartinSmeyer ?

MartinSmeyer commented:
> There still is an issue with the calculation of the visibility masks which happens based on depth image comparisons ([code](https://github.com/wboerdijk/BlenderProc/blob/c5607a83b290769b7b8099f0be499590fc1cff66/blenderproc/python/writer/BopWriterUtility.py#L577-L591)), where there are occasionally small pixel-wise artifacts (also visible in the rendered images, e.g. [this one](https://camo.githubusercontent.com/b9fab24ff62d10355e2b33bea4e63c1091046163ebab60fa3b8db048ee397f30/68747470733a2f2f6c6976652e737461746963666c69636b722e636f6d2f36353533352f35323836363732353431345f323966623332323039362e6a7067)). This only happens at borders of objects, and should be due to different alias handling in blender and pyrender I assume? I tried playing around with disabling anti-aliasing (e.g. [this suggestion](https://github.com/mmatl/pyrender/pull/175/files)), but similar artifacts remain. Maybe you have any additional suggestions @MartinSmeyer ?

Yes, both pyrender and Blender have anti-aliasing turned on during depth rendering. The results still differ because anti-aliasing can be performed with various different settings and techniques. I think the visibility fraction shouldn't change too much and was likely not exact with vispy depth either. It's mainly used for the evaluation on real scenes, so don't worry. Thanks for testing the differences though, it's interesting.

MartinSmeyer commented on blenderproc/python/types/MeshObjectUtility.py:
```diff
@@ -522,6 +523,20 @@ def add_modifier(self, name: str, **kwargs):
         for key, value in kwargs.items():
             setattr(modifier, key, value)
 
+    def mesh_as_trimesh(self) -> Trimesh:
```
To avoid this duplicate function, you could just add a non-member function to BlenderUtility, e.g. 

```python
def mesh_as_trimesh(entity: Entity) -> Optional[trimesh]
   if isinstance(entity, Link):
   ...
   elif isinstance(entity, MeshObject):
   ...
   else:
   "Only meshobjects and Links are currently supported"
```

cornerfarmer replied:
I think that is not necessary as `self.visuals[0]` of `Link` is a `MeshObject`, so you could just replace the duplicate code in `Link` with:
```
        return self.visuals[0].mesh_as_trimesh()
```

MartinSmeyer replied:
okay fine as well, just wanted to avoid having two functions that do the same :)

wboerdijk replied:
true, thanks! (addressed https://github.com/DLR-RM/BlenderProc/pull/881/commits/691b4423b7feda9979e0baf1268225b6969fbdaa)

MartinSmeyer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -131,14 +133,46 @@ def write_bop(output_dir: str, target_objects: Optional[List[MeshObject]] = None
         chunk_dirs = chunk_dirs[chunk_dir_ids.index(f"{starting_chunk_id:06d}"):]
 
+        # convert all objects to trimesh objects
+        trimesh_objects = {}
+        for obj in dataset_objects:
+            if isinstance(obj, Link):
+                if not obj.visuals:
+                    continue
+                if len(obj.visuals) > 1:
+                    warnings.warn('BOP Writer only supports saving annotations of one visual mesh per Link')
+            trimesh_obj = obj.mesh_as_trimesh()
+            # we need to create a double-sided material to be able to render non-watertight meshes
+            # the other parameters are defaults, see
+            # https://github.com/mmatl/pyrender/blob/master/pyrender/mesh.py#L216-L223
+            material = pyrender.MetallicRoughnessMaterial(alphaMode='BLEND', baseColorFactor=[0.3, 0.3, 0.3, 1.0],
+                                                          metallicFactor=0.2, roughnessFactor=0.8, doubleSided=True)
+            trimesh_objects[obj.get_cp('category_id')] = pyrender.Mesh.from_trimesh(mesh=trimesh_obj, material=material)
```
This loop will run unnecessarily often if you have many objects of the same `category_id` in the scene. So would run a check in line 139:

```python
if obj.get_cp('category_id') in trimesh_objects:
   continue
```

wboerdijk replied:
yes, but that would not allow a user with duplicate objects in the scene to get bop annotations from them. is that desired?

MartinSmeyer replied:
I think this would still work as only one mesh is rendered per scene anyways? Currently you just overwrite the trimesh_objects dict entry in every iteration if there is a duplicate object.

wboerdijk replied:
oh yes there you have a point, forgot about that. sure, i'll adapt the loop accordingly!

MartinSmeyer commented:
Minor comments, great feature! :+1:

cornerfarmer commented:
BlenderProcHelper result for run 501:

BlenderProcHelper has started, waiting for first results.

cornerfarmer commented:
@wboerdijk Thanks, looks good! Could you please resolve the pylint issues, then we can merge the PR

wboerdijk commented:
@cornerfarmer sorry, completely overlooked this! That's because of the headless rendering; it needs to be set before importing pyrender (https://pyrender.readthedocs.io/en/latest/examples/offscreen.html). I disabled the respective linting there!

cornerfarmer commented:
BlenderProcHelper result for run 502:

BlenderProcHelper has started, waiting for first results.
